File: week_2/go/task_d/task_d.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	s, _ := reader.ReadString('\n')
	s = strings.TrimSpace(s)

	var n int
	fmt.Fscan(reader, &n)

	dict := make(map[string]int)
	for i := 0; i < n; i++ {
		var temp string
		fmt.Fscan(reader, &temp)
		dict[temp]++
	}

	l := len(s)
	dp := make([]bool, l+1)
	prev := make([]int, l+1)
	dp[0] = true

	for i := 1; i <= l; i++ {
		prev[i] = -1
		for j := 0; j < i; j++ {
			if dp[j] {
				cur := s[j:i]
				_, exists := dict[cur]
				if exists {
					dp[i] = true
					prev[i] = j
					break
				}

			}
		}
	}

	pos := l
	var res []string

	for pos > 0 {
		beg := prev[pos]
		if beg >= 0 {
			res = append(res, s[beg:pos])
		}
		pos = beg
	}

	for i := range res {
		r_ind := len(res) - 1 - i
		fmt.Fprint(writer, res[r_ind], " ")
	}

}

File: week_2/task_d.cc
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

int main() {
  std::string s;
  int n = 0;

  std::cin >> s;
  std::cin >> n;
  std::unordered_map<std::string, int> dict;
  for (int i = 0; i < n; ++i) {
    std::string temp;
    std::cin >> temp;
    ++dict[temp];
  }

  int l = s.size();
  std::vector<bool> dp(l + 1, false);
  std::vector<int> prev(l + 1, -1);
  dp[0] = true;

  for (int i = 1; i <= l; ++i) {
    for (int j = 0; j < i; ++j) {
      if (dp[j]) {
        std::string cur;
        for (int k = j; k < i; ++k) {
          cur += s[k];
        }
        if (dict.contains(cur)) {
          dp[i] = true;
          prev[i] = j;
          break;
        }
      }
    }
  }

  int pos = l;
  std::stack<std::string> res;
  while (pos > 0) {
    int beg = prev[pos];
    std::string cur;
    for (int i = beg; i < pos; ++i) {
      cur += s[i];
    }
    res.push(cur);
    pos = beg;
  }

  while (!res.empty()) {
    std::cout << res.top() << ' ';
    res.pop();
  }

  return 0;
}